lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ActiveContextManager.cs

```diff
@@ -0,0 +1,154 @@
+        /// Checks if all the requested contexts are received. 
+        /// </summary>
+        internal bool AllContextsReceived
+        {
+            get { return NumberOfMissingContexts == 0; }
+        }
+
+        /// <summary>
+        /// Adding an IActiveContext to the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext already exists.
+        /// </summary>
+        /// <param name="activeContext"></param>
+        internal void Add(IActiveContext activeContext)
+        {
+            if (_activeContexts.ContainsKey(activeContext.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] received already exists.", activeContext.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (NumberOfActiveContexts >= _totalExpectedContexts)
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "Trying to add an extra active context {0}. The total number of the active contexts has reached to the expected number {1}.", activeContext.Id, _totalExpectedContexts);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _activeContexts.Add(activeContext.Id, activeContext);
+        }
+
+        /// <summary>
+        /// Remove an IActiveContext from the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext doesn't exist.
+        /// </summary>
+        /// <param name="activeContextId"></param>
+        internal void Remove(string activeContextId)
+        {
+            if (!_activeContexts.ContainsKey(activeContextId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] to be removed does not exist.", activeContextId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+            _activeContexts.Remove(activeContextId);
+        }
+
+        /// <summary>
+        /// Returns the current number of IActiveContext in the ActiveContext collection
+        /// </summary>
+        internal int NumberOfActiveContexts
+        {
+            get { return _activeContexts.Count; }
+        }
+
+        /// <summary>
+        /// Given an IFailedEvaluator, remove associated IActiveContext from the collection
+        /// Throw SystemException if associated IActiveContext doesn't exist.
+        /// Throw IMRUSystemException if more than one IActiveContexts are associated with the IFailedEvaluator
+        /// as current IMRU driver assumes that there is only one context associated with the IFailedEvalutor
+        /// </summary>
+        /// <param name="value"></param>
+        internal void RemoveFailedContextInFailedEvaluator(IFailedEvaluator value)
+        {
+            if (value.FailedContexts != null && value.FailedContexts.Count > 0)
+            {
+                if (value.FailedContexts.Count == 1)
```

Comment:
In IMRU, we assume one context on one evalutor and submit task on each active context. We can make ActiveContextManager more generic, but in IMRU driver, without this assumption, task submission has to be handled differently.